src/Compilers/CSharp/Test/Emit/CodeGen/SwitchTests.cs

```diff
@@ -7612,5 +7612,73 @@ .maxstack  1
         }
 
         #endregion
+
+        #region regression tests
+
+        [Fact, WorkItem(18859, "https://github.com/dotnet/roslyn/issues/18859")]
+        public void BoxInPatternSwitch_01()
+        {
+            var source = @"using System;
+
+public class Program
+{
+    public static void Main()
+    {
+        switch (StringSplitOptions.RemoveEmptyEntries)
+        {
+            case object o:
+                Console.WriteLine(o);
+                break;
+        }
+    }
+}";
+            var compVerifier = CompileAndVerify(source, expectedOutput: "RemoveEmptyEntries");
+            compVerifier.VerifyIL("Program.Main",
+@"{
+  // Code size       12 (0xc)
+  .maxstack  1
+  IL_0000:  ldc.i4.1
+  IL_0001:  box        ""System.StringSplitOptions""
+  IL_0006:  call       ""void System.Console.WriteLine(object)""
+  IL_000b:  ret
+}"
+            );
+        }
+
+        [Fact, WorkItem(18859, "https://github.com/dotnet/roslyn/issues/18859")]
+        public void BoxInPatternIf_02()
+        {
+            var source = @"using System;
+
+public class Program
+{
+    public static void Main()
+    {
+        if (StringSplitOptions.RemoveEmptyEntries is object o)
+        {
+            Console.WriteLine(o);
+        }
+    }
+}";
+            var compVerifier = CompileAndVerify(source, expectedOutput: "RemoveEmptyEntries");
+            compVerifier.VerifyIL("Program.Main",
+@"{
+  // Code size       22 (0x16)
+  .maxstack  2
+  .locals init (object V_0) //o
+  IL_0000:  ldc.i4.1
+  IL_0001:  box        ""System.StringSplitOptions""
+  IL_0006:  isinst     ""object""
```

Comment:
We do perform quite a few optimizations under `/optimize`. Evidently the `ReleaseDll` below folded the redundant computation of the bool result of the null-check on isinst result into simpler "isnull" branch.

We are not optimizing  "isinst object" since is not generally optimizable. The value could be null. Even if the argument is a `box` instruction.
We can dig into the arguments and check whether the argument is a box of a non-nullable value type. This combination was not common (or possible?) until this feature.

If you think `is object o` pattern will be common enough in actual user code, then we need to track this as a "Code Quality Improvement" bug. There is a tag for that.

See: https://github.com/dotnet/roslyn/labels/Concept-Code%20Quality%20Improvement